libcnb-cargo/src/cross_compile.rs

```diff
@@ -0,0 +1,91 @@
+/// platform to the desired target platform.
+///
+/// This function will not install required toolchains, linkers or compilers automatically. It will
+/// look for the required tools and returns an error if they cannot be found. See
+/// [`cross_compile_help`] on how to support users setting up their machine for cross-compilation.
+///
+/// # Errors
+///
+/// Will return `Err` if the cross-compile environment could not be created for the current host
+/// and target platform.
+pub fn cross_compile_env(
+    target_triple: impl AsRef<str>,
+) -> Result<Vec<(OsString, OsString)>, CrossCompileError> {
+    // Background: https://omarkhawaja.com/cross-compiling-rust-from-macos-to-linux/
+    if target_triple.as_ref() == X86_64_UNKNOWN_LINUX_MUSL && cfg!(target_os = "macos") {
+        let gcc_binary_name = "x86_64-linux-musl-gcc";
+
+        which(gcc_binary_name)
+            .map(|gcc_binary_path| {
+                vec![
+                    (
+                        // Required until Cargo can auto-detect the musl-cross gcc/linker itself,
+                        // since otherwise it checks for a binary named 'musl-gcc'
+                        // not 'x86_64-linux-musl-gcc':
+                        // https://github.com/FiloSottile/homebrew-musl-cross/issues/16
+                        // https://github.com/rust-lang/cargo/issues/4133
+                        OsString::from("CARGO_TARGET_X86_64_UNKNOWN_LINUX_MUSL_LINKER"),
+                        OsString::from(&gcc_binary_path),
+                    ),
+                    (
+                        // Required so that any crates that call out to gcc are also cross-compiled:
+                        // https://github.com/alexcrichton/cc-rs/issues/82
+                        OsString::from("CC_x86_64_unknown_linux_musl"),
+                        OsString::from(&gcc_binary_path),
+                    ),
+                ]
+            })
+            .map_err(|_| {
+                CrossCompileError::CouldNotFindRequiredBinary(String::from(gcc_binary_name))
+            })
+    } else {
+        Ok(vec![])
+    }
+}
+
+#[derive(Debug)]
+pub enum CrossCompileError {
+    CouldNotFindRequiredBinary(String),
+}
+
+/// Returns a human-readable help text about cross-compiling from the user's host platform to
+/// the desired target platform.
+pub fn cross_compile_help(target_triple: impl AsRef<str>) -> Option<String> {
+    let mut help_texts = HashMap::new();
+
+    help_texts.insert(
+        ("macos", X86_64_UNKNOWN_LINUX_MUSL),
+        String::from(
+            r#"For cross-compilation from macOS to x86_64-unknown-linux-musl, a C compiler and
+linker for the target platform must be installed on your computer.
+
+The easiest way to install 'x86_64-linux-musl-gcc' is to follow the instructions in the linked
+GitHub repository:
+
+https://github.com/FiloSottile/homebrew-musl-cross"#,
+        ),
+    );
+
+    help_texts.insert(
+        ("linux", X86_64_UNKNOWN_LINUX_MUSL),
```

Comment:
@edmorley I have this in #214, maybe we can punt on improving it there?

Reply:
I refactored the way the cross-compilation functions work. With Linux, we don't need to set environment variables, but we want to check for binaries. This made both the naming and design of `cross_compile_env` not a good fit. It was replaced by a `cross_compile_assistance` function which can much easier extended in the future.

Also, splitting the help text from the actual checks was clunky. `cross_compile_assistance` now returns the help text instead of an error if the cross-compile settings could not be determined. This removes the burden from the call-site to first detect the environment could not be set and then display the help text. With this change, the help text can change when the issue changes. Right now we only have single checks, but it will be useful when we have to do more.

The change also made it clearer that it is not an error if there are no cross-compile settings for a target platform as this is just an heuristic at best.

Oh, and the Linux help text is now displayed if `musl-gcc` is missing. ;-)

See: fb16494